File: cpp/ros2/src/sando_cpp/src/convert_goal_to_cmd_vel.cpp
// convert_goal_to_cmd_vel.cpp — C++ ROS2 port of sando_py/nodes/convert_goal_to_cmd_vel.py.
// Lyapunov tracking controller turning a flat Goal into a ground-robot Twist (cmd_vel):
//   v_cmd    = v_d*cos(eyaw) - kx*ex
//   yawd_cmd = dyaw_d - v_d*(ky*ey + sin(eyaw))/sqrt(ey^2+eps^2) - kyaw*eyaw
// where (ex,ey) is the position error rotated into the desired-yaw frame. Control law
// identical to the Python (and the original C++ baseline).
#include <rclcpp/rclcpp.hpp>
#include <geometry_msgs/msg/twist.hpp>
#include <dynus_interfaces/msg/goal.hpp>
#include <dynus_interfaces/msg/state.hpp>
#include <array>
#include <cmath>
#include <chrono>
using namespace std::chrono_literals;

static inline double yaw_from_quat(double qx, double qy, double qz, double qw) {
  double siny = 2.0 * (qw * qz + qx * qy);
  double cosy = 1.0 - 2.0 * (qy * qy + qz * qz);
  return std::atan2(siny, cosy);
}

class GoalToCmdVel : public rclcpp::Node {
 public:
  GoalToCmdVel() : rclcpp::Node("goal_to_cmd_vel") {
    state_pos_ = {declare_parameter<double>("x", 0.0), declare_parameter<double>("y", 0.0),
                  declare_parameter<double>("z", 0.0)};
    state_yaw_ = declare_parameter<double>("yaw", 0.0);
    std::string cmd_topic = declare_parameter<std::string>("cmd_vel_topic_name", "cmd_vel");
    kx_ = declare_parameter<double>("ground_robot_kx", 1.0);
    ky_ = declare_parameter<double>("ground_robot_ky", 1.0);
    kyaw_ = declare_parameter<double>("ground_robot_kyaw", 1.0);
    eps_ = declare_parameter<double>("ground_robot_eps", 1e-2);
    RCLCPP_INFO(get_logger(), "kx=%g, ky=%g, kyaw=%g, eps=%g", kx_, ky_, kyaw_, eps_);

    pub_ = create_publisher<geometry_msgs::msg::Twist>(cmd_topic, 10);
    sub_goal_ = create_subscription<dynus_interfaces::msg::Goal>(
        "goal", 10, std::bind(&GoalToCmdVel::goal_cb, this, std::placeholders::_1));
    sub_state_ = create_subscription<dynus_interfaces::msg::State>(
        "state", 10, std::bind(&GoalToCmdVel::state_cb, this, std::placeholders::_1));
    timer_ = create_wall_timer(10ms, std::bind(&GoalToCmdVel::tick, this));
  }

 private:
  void state_cb(const dynus_interfaces::msg::State::SharedPtr m) {
    state_pos_ = {m->pos.x, m->pos.y, m->pos.z};
    state_yaw_ = yaw_from_quat(m->quat.x, m->quat.y, m->quat.z, m->quat.w);
    state_init_ = true;
  }
  void goal_cb(const dynus_interfaces::msg::Goal::SharedPtr m) {
    goal_p_ = {m->p.x, m->p.y, m->p.z};
    goal_v_ = {m->v.x, m->v.y, m->v.z};
    goal_yaw_ = m->yaw; goal_dyaw_ = m->dyaw;
    goal_init_ = true;
  }
  void tick() {
    if (!state_init_ || !goal_init_) return;
    double x_d = goal_p_[0], y_d = goal_p_[1];
    double v_d = std::hypot(goal_v_[0], goal_v_[1]);
    double sx = state_pos_[0], sy = state_pos_[1];
    double ex = std::cos(goal_yaw_) * (sx - x_d) + std::sin(goal_yaw_) * (sy - y_d);
    double ey = -std::sin(goal_yaw_) * (sx - x_d) + std::cos(goal_yaw_) * (sy - y_d);
    double eyaw = state_yaw_ - goal_yaw_;
    double v_cmd = v_d * std::cos(eyaw) - kx_ * ex;
    double denom = std::sqrt(ey * ey + eps_ * eps_);
    double yawd_cmd = goal_dyaw_ - v_d * (ky_ * ey + std::sin(eyaw)) / denom - kyaw_ * eyaw;
    geometry_msgs::msg::Twist t;
    t.linear.x = v_cmd; t.angular.z = yawd_cmd;
    pub_->publish(t);
  }
  std::array<double, 3> state_pos_{0, 0, 0}, goal_p_{0, 0, 0}, goal_v_{0, 0, 0};
  double state_yaw_ = 0, goal_yaw_ = 0, goal_dyaw_ = 0;
  double kx_, ky_, kyaw_, eps_;
  bool state_init_ = false, goal_init_ = false;
  rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr pub_;
  rclcpp::Subscription<dynus_interfaces::msg::Goal>::SharedPtr sub_goal_;
  rclcpp::Subscription<dynus_interfaces::msg::State>::SharedPtr sub_state_;
  rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char** argv) {
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<GoalToCmdVel>());
  rclcpp::shutdown();
  return 0;
}

File: python/sando_py/nodes/convert_goal_to_cmd_vel.py
"""Goal -> cmd_vel for ground robots — Python port of convert_goal_to_cmd_vel.cpp.

Implements the same Lyapunov-based controller as the C++ node:
  v_command   = v_desired * cos(eyaw) - kx * ex
  yawd_command = yawd_desired
                  - v_desired * (ky*ey + sin(eyaw)) / sqrt(ey^2 + eps^2)
                  - kyaw * eyaw
where (ex, ey) is the position error rotated into the desired yaw frame.

中文说明:
给「地面机器人」用的桥接 + 控制节点。规划器输出的是飞机式的 Goal(期望位置/速度/偏航),
但差速底盘只能接收「前进线速度 + 转向角速度」(geometry_msgs/Twist,话题 cmd_vel)。
这个节点把 Goal 转成底盘能执行的 cmd_vel。

它用的是一个基于李雅普诺夫(Lyapunov)思想设计的轨迹跟踪控制器——大白话就是:
比较「当前位姿」和「期望位姿」的差,根据这个误差算出该给多大的前进速度和转向速度,
让机器人逐渐贴回期望轨迹。这是经典的非完整移动机器人(只能前进+转向)跟踪控制律。

输入:话题 `goal`(期望状态) + 话题 `state`(机器人当前状态)。
输出:话题 `cmd_vel`(底盘的速度指令)。控制律和 C++ 版完全一致。
"""
from __future__ import annotations

import math

import rclpy
from rclpy.node import Node
from geometry_msgs.msg import Twist

from ..utils import yaw_from_quat


# 延迟导入 Goal 消息类型(没装就返回 None)
def _get_goal_msg():
    try:
        from dynus_interfaces.msg import Goal
        return Goal
    except ImportError:
        return None


# 延迟导入 State 消息类型(没装就返回 None)
def _get_state_msg():
    try:
        from dynus_interfaces.msg import State
        return State
    except ImportError:
        return None


class GoalToCmdVel(Node):
    """把规划器的 Goal 转成差速底盘的 cmd_vel(线速度+角速度)的跟踪控制节点。"""
    def __init__(self):
        super().__init__("goal_to_cmd_vel")
        self.current_yaw = 0.0
        # 两个就绪标志:state 和 goal 都收到过才开始算控制量
        self.state_initialized = False
        self.goal_initialized = False

        x = self.declare_parameter("x", 0.0).value
        y = self.declare_parameter("y", 0.0).value
        z = self.declare_parameter("z", 0.0).value
        yaw = self.declare_parameter("yaw", 0.0).value
        cmd_vel_topic = self.declare_parameter("cmd_vel_topic_name", "cmd_vel").value
        # kx/ky/kyaw 是控制增益(误差放大系数):越大纠偏越猛但越容易抖
        self.kx = float(self.declare_parameter("ground_robot_kx", 1.0).value)
        self.ky = float(self.declare_parameter("ground_robot_ky", 1.0).value)
        self.kyaw = float(self.declare_parameter("ground_robot_kyaw", 1.0).value)
        # eps:防止后面公式分母 sqrt(ey^2+eps^2) 为零的小正数
        self.eps = float(self.declare_parameter("ground_robot_eps", 1e-2).value)

        self.state_pos = (float(x), float(y), float(z))
        self.state_yaw = float(yaw)
        self.goal_p = (0.0, 0.0, 0.0)
        self.goal_v = (0.0, 0.0, 0.0)
        self.goal_yaw = 0.0
        self.goal_dyaw = 0.0

        self.get_logger().info(
            f"kx={self.kx}, ky={self.ky}, kyaw={self.kyaw}, eps={self.eps}"
        )

        self.pub_cmd_vel = self.create_publisher(Twist, cmd_vel_topic, 10)
        GoalMsg = _get_goal_msg()
        StateMsg = _get_state_msg()
        if GoalMsg is not None:
            self.sub_goal = self.create_subscription(GoalMsg, "goal", self._goal_cb, 10)
        if StateMsg is not None:
            self.sub_state = self.create_subscription(StateMsg, "state", self._state_cb, 10)
        # 100Hz 控制循环
        self.timer = self.create_timer(0.01, self._tick)

    # 收到当前状态:记下位置,并把四元数转成平面偏航角 yaw
    def _state_cb(self, msg) -> None:
        self.state_pos = (msg.pos.x, msg.pos.y, msg.pos.z)
        self.state_yaw = yaw_from_quat(msg.quat.x, msg.quat.y, msg.quat.z, msg.quat.w)
        self.state_initialized = True

    # 收到期望状态:记下期望位置/速度/偏航/偏航角速度
    def _goal_cb(self, msg) -> None:
        self.goal_p = (msg.p.x, msg.p.y, msg.p.z)
        self.goal_v = (msg.v.x, msg.v.y, msg.v.z)
        self.goal_yaw = float(msg.yaw)
        self.goal_dyaw = float(msg.dyaw)
        self.goal_initialized = True

    # 100Hz:根据当前与期望的差算出 cmd_vel(控制律见文件头公式)
    def _tick(self) -> None:
        if not self.state_initialized or not self.goal_initialized:
            return
        x_d, y_d, _ = self.goal_p
        xd_d, yd_d, _ = self.goal_v
        yaw_d, dyaw_d = self.goal_yaw, self.goal_dyaw
        # v_d:期望前进速率(水平速度的模长)
        v_d = math.hypot(xd_d, yd_d)
        sx, sy, _ = self.state_pos
        # 把位置误差旋转到「期望朝向」坐标系:ex=纵向(前后)误差,ey=横向(左右)误差
        ex = math.cos(yaw_d) * (sx - x_d) + math.sin(yaw_d) * (sy - y_d)
        ey = -math.sin(yaw_d) * (sx - x_d) + math.cos(yaw_d) * (sy - y_d)
        # eyaw:朝向误差
        eyaw = self.state_yaw - yaw_d
        # 线速度指令:前馈期望速度,减去纵向误差的纠偏项
        v_command = v_d * math.cos(eyaw) - self.kx * ex
        # 分母加 eps 防止除零(eps 在构造里声明)
        denom = math.sqrt(ey * ey + self.eps * self.eps)
        # 角速度指令:前馈期望角速度,再叠加横向误差和朝向误差的纠偏项
        yawd_command = dyaw_d - v_d * (self.ky * ey + math.sin(eyaw)) / denom - self.kyaw * eyaw
        twist = Twist()
        twist.linear.x = float(v_command)
        twist.angular.z = float(yawd_command)
        self.pub_cmd_vel.publish(twist)


def main(args=None):
    rclpy.init(args=args)
    node = GoalToCmdVel()
    try:
        rclpy.spin(node)
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == "__main__":
    main()
